File: Difficulty: Medium/Minimal Cost/minimal-cost.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    int minimizeCost(int k, vector<int>& arr) {
        // Code here
           int n = arr.size();
        vector<int> dp(n, INT_MAX); 
        dp[0] = 0; 

        for (int i = 0; i < n; ++i) {
            for (int jump = 1; jump <= k && i + jump < n; ++jump) {
                dp[i + jump] = min(dp[i + jump], dp[i] + abs(arr[i] - arr[i + jump]));
            }
        }

        return dp[n - 1];  // Return the minimum cost to reach the last index
    }
};

//{ Driver Code Starts.

int main() {
    string ts;
    getline(cin, ts);
    int t = stoi(ts);
    while (t--) {
        string ks;
        getline(cin, ks);
        int k = stoi(ks);
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        int res = obj.minimizeCost(k, arr);
        cout << res << endl;
        // string tl;
        // getline(cin, tl);
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Minimal Cost/minimal-cost.java
//{ Driver Code Starts
// Initial Template for Java
import java.io.*;
import java.lang.*;
import java.util.*;


// } Driver Code Ends
// User function Template for Java

class Solution {
    
    public int fun(int ind, int k, int[] arr,int[] dp)
    {
        if(ind == 0)
        {
            return 0;
        }
        
        if(dp[ind] != -1) return dp[ind];
        
        int result = Integer.MAX_VALUE;
        
        for(int val = 1; val <= k; val++)
        {
            int energy = Integer.MAX_VALUE;
            
            if(ind - val >= 0)
            {
                energy = Math.abs(arr[ind] - arr[ind-val]) + fun(ind - val,k,arr,dp);
            }
            
            result = Math.min(energy,result);
        }
        
        return dp[ind] = result;
    }
    
    public int minimizeCost(int k, int arr[]) {
        // code here
        int n = arr.length;
        int[] dp = new int[n];
        Arrays.fill(dp,-1);
        return fun(n-1,k,arr,dp);
    }
}


//{ Driver Code Starts.

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            int k = Integer.parseInt(br.readLine());
            String line = br.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;
            Solution obj = new Solution();
            int res = obj.minimizeCost(k, arr);

            System.out.println(res);

            System.out.println("~");
        }
    }
}
// } Driver Code Ends